dandi/files.py

```diff
@@ -838,64 +836,98 @@ def iter_upload(
             )
         a = RemoteAsset.from_data(dandiset, r)
         assert isinstance(a, RemoteZarrAsset)
-        to_upload: List[dict] = []
-        to_delete: List[RemoteZarrEntry] = []
-        for p in stat.files:
-            pdigest = p.get_digest().value
-            item = {"path": str(p), "etag": pdigest}
-            try:
-                e = old_zarr_entries.pop(str(p))
-            except KeyError:
-                for pp in p.parents:
-                    pps = str(pp)
-                    if pps in old_zarr_entries:
-                        if old_zarr_entries[pps].is_file():
+
+        to_upload = EntryUploadTracker()
+        if old_zarr_entries:
+            to_delete: List[RemoteZarrEntry] = []
+            digesting: List[Future[Optional[Tuple[LocalZarrEntry, str]]]] = []
+            yield {"status": "comparing against remote Zarr"}
+            with ThreadPoolExecutor(max_workers=jobs or 5) as executor:
```

Comment:
did you try this PR on some size-able zarr upload from e.g. hub to our staging S3?  I wonder about how long it took e.g. to start with uploading for a fresh new zarr (so nothing on server yet) and for e.g. some interrupted upload after some files were already transmitted; and what speed did you observe?

Reply:
Not yet.  Do you have a Zarr or script for creating one to recommend using, or should I just jump straight to trying to upload `test128.ngff`?  (CC: @satra)

Reply:
I have also downloaded that sample large-ish  (~37GB) zarr @satra uploaded to staging , into the hub: `/shared/zarr/b35681cd-936a-477c-ac51-06661612d4df` . Would be interesting to gauge upload speed for that one into staging (as a new zarr of cause), and then interrupt and continue.  @satra also placed a helper tool (like dstat) under `/shared/io-utils/dool` to monitor "wire bandwidth" (in addition to what client would be reporting)

Reply:
@yarikoptic While testing out the upload, I found an inefficiency: the first thing `upload()` does when processing an asset is get its local size, which for Zarrs means iterating over the entire directory tree, which is later iterated over again inside `iter_upload()`.  Should Zarr uploads delay this calculation until the latter point?

Reply:
@yarikoptic I just finished a complete upload of the given Zarr in a single upload session, which took 63 minutes, the first couple minutes of which were spent in pre-upload directory traversals.  The upload itself alternated between uploading files in fast bursts of about 10 seconds each and waiting about 10-20 seconds for `/zarr/{zarr_id}/upload/complete/` requests to return.  Next I'll try interrupting and resuming.

Reply:
re size compute iteration: I guess iteration through zarr to determine the size might just need to be improved (that threaded walk?) since probably otherwise would be too tricky for "compute out of band and update progress reporting etc accordingly later". Or you see how to make it delayed "nicely"?

So -- we still have some dandi-api activities (I guess checksumming) to return for `/complete` which should be either optimized or somehow shifted to be "out of band" besides may be the final batch upload?  (attn @satra @dchiquito @waxlamp )

Reply:
@yarikoptic By "delay", I meant leaving the display of the "SIZE" column for a Zarr blank until all the local entries have been traversed while determining what to upload, at which point (just before starting upload proper) `iter_upload()` can just emit a `{"size": total_size}` dict, the size having been calculated while iterating.

Reply:
if this is simply for the metadata, i would say the client could provide it but there is no reason for server to trust it. and since the server will compute size and checksum out of band, the server will populate those fields in the metadata. this is similar to the PR that @AlmightyYakob is working on that will also support out of band ingestion.

Reply:
@satra This is not for the metadata; this is for the pyout display.

Reply:
ah in that case, i would say don't worry about total size. just provide a summary at the end.

Reply:
> @yarikoptic By "delay", I meant leaving the display of the "SIZE" column for a Zarr blank until all the local entries have been traversed while determining what to upload, at which point (just before starting upload proper) `iter_upload()` can just emit a `{"size": total_size}` dict, the size having been calculated while iterating.

that sounds great to me.

Reply:
@satra - don't worry about size.  But we need to figure out what to do with `/complete` for finishing a batch upload for zarr which seems to add ~200% overhead on top of transfer time in @jwodder's experiment.